### PR TITLE
fix(next_core): do not apply user config to internal assets

### PR DESCRIPTION
**Pull request by kwonoj**
### What?
This is necessary changes for the WEB-862, however dependent to https://github.com/vercel/turbo/pull/4521.

### CODE REVIEW

File: packages/next-swc/crates/next-core/src/next_client/context.rs [hunk 1]
```diff
@@ -204,10 +207,22 @@ pub async fn get_client_module_options_context(
...
-            foreign_code_context_condition(next_config).await?,
-            module_options_context.clone().cell(),
-        )],
+        rules: vec![
+            (
+                foreign_code_context_condition(next_config).await?,
+                module_options_context.clone().cell(),
+            ),
+            // If module is internal assets (i.e overlay, fallback) coming from embed_js fs root
+            // We will not apply any user defined transforms
```
Comment (alexkirsz) [inline packages/next-swc/crates/next-core/src/next_client/context.rs:216]:
```suggestion
            // If the module is an internal asset (i.e overlay, fallback) coming from the embedded FS,
            // don't apply user defined transforms.
```

File: packages/next-swc/crates/next-core/src/next_client/context.rs [hunk 2]
```diff
@@ -204,10 +207,22 @@ pub async fn get_client_module_options_context(
...
+            (
+                ContextCondition::InPath(next_js_fs().root()),
+                (ModuleOptionsContext {
+                    enable_typescript_transform: Some(TypescriptTransformOptions::default().cell()),
+                    ..module_options_context.clone()
+                })
```
Comment (alexkirsz) [inline packages/next-swc/crates/next-core/src/next_client/context.rs:222]:
```suggestion
                ModuleOptionsContext {
                    enable_typescript_transform: Some(TypescriptTransformOptions::default().cell()),
                    ..module_options_context.clone()
                }
```

**Review by alexkirsz — COMMENTED** [review]
Can we test this somehow? The way it's written now, I would expect us to break this unexpectedly in the future.

**Comment by kwonoj** [comment]
> Can we test this somehow? The way it's written now, I would expect us to break this unexpectedly in the future.

Mind share the scope we want to test? For the happy path (that user having offending config) there are tests like https://github.com/vercel/next.js/blob/canary/test/development/basic/emotion-swc/jsconfig.json#L3 confirms behavior. Otherwise for the negative path (if internal asset fails to render) it caused hard fail so any of integration test would fail.

File: packages/next-swc/crates/next-core/src/next_client/context.rs [hunk 3]
```diff
@@ -205,10 +208,22 @@ pub async fn get_client_module_options_context(
...
-        )],
+        rules: vec![
+            (
+                foreign_code_context_condition(next_config).await?,
+                module_options_context.clone().cell(),
+            ),
+            // If the module is an internal asset (i.e overlay, fallback) coming from the embedded FS,
+            // don't apply user defined transforms.
+            (
+                ContextCondition::InPath(next_js_fs().root()),
```
Comment (sokra) [inline packages/next-swc/crates/next-core/src/next_client/context.rs:219]: There are more paths that are affected. `turbopack-dev` has an filesystem too

**Review by sokra — CHANGES_REQUESTED** [review]
Could you add a test case for the bug

**Comment by kwonoj** [comment]
> Could you add a test case for the bug

@sokra As wrote above, I'm thinking existing integration test (emotion) could be a happy path verification for this instead of making some duplicated test under next-dev-tests. Thoughts? or we should duplicate?

**Comment by sokra** [comment]
> I'm thinking existing integration test (emotion) could be a happy path verification for this instead of making some duplicated test under [next-dev-tests](https://linear.app/vercel/issue/NEXT-dev-tests).

I'm fine with that. Does they pass for that PR? Where can I see that?

I think it shouldn't pass as there is the filesystem for turbopack-dev missing. So I wonder if it's passing?

**Comment by kwonoj** [comment]
> Does they pass for that PR? Where can I see that?

It passed locally, so I'm planning to prep PR like this https://github.com/vercel/turbo/pull/4610 after landing this.

> I think it shouldn't pass as there is the filesystem for turbopack-dev missing. So I wonder if it's passing?

No, it passed. I believe that's cause of user config that affects the internal output (`jsxImportSource`) is not being used by any assets under `turbopack-dev` anyway.

**Comment by sokra** [comment]
> Regardless, my question still remains: shouldn't turbopack-* need to control how turbopack-dev being transformed? otherwise say any consumer other than [next-dev](https://linear.app/vercel/issue/NEXT-dev) in the future they can control (and break) how those assets being compiled by their own configs.

probably yes.

> jsxImportSource

I would also say a tsconfig.json should only affect files in that directory/package and doesn't have global effects.

e.g. in a monorepo you could have multiple packages with different tsconfigs.json which have different jsxImportSources.

**Comment by kwonoj** [comment]
> probably yes.

Yes, I'm trying to make a PR to turbopack to this.

> I would also say a tsconfig.json should only affect files in that directory/package and doesn't have global effects.

Yes, agreed too. What should be changed in this PR?

**Comment by sokra** [comment]
I think there are two types or configuration that should apply differently.

* global configuration that applies to all modules (potentially bringing their own filtering)
    * `ModuleOptionContext`
    * `ResolveOptionContext`
* local configuration that only applies to a certain context (usually a directory)
    * `tsconfig.json`
    * `package.json` (e. g. `browser` alias field, `sideEffects: false` flag, `type: "module"`

**Comment by kwonoj** [comment]
It sounds like it's another changes we want to how to deal with configs - while this PR specifically aims for `excluding` any user level configs for the internal assets. Am I reading something incorrect that this PR should include all of suggested changes?

**Comment by sokra** [comment]
> Am I reading something incorrect that this PR should include all of suggested changes?

Let's make that a follow-up PR

**Review by sokra — APPROVED** [review]

**Review by jridgewell — DISMISSED** [review]
Alternatively, we can pass the `Asset` instead of the `FileSystemPathVc`, which would allow us to detect that the asset is a virtual one (either our internal assets, or a asset we dynamically construct).

**Comment by sokra** [comment]
> which would allow us to detect that the asset is a virtual one (either our internal assets, or a asset we dynamically construct).

We had this before, but that's not that great since in future other plugins could use virtual assets too and they shouldn't be affected.

**Review by sokra — APPROVED** [review]

**Review by sokra — DISMISSED** [review]

**Comment by ijjk** [comment]
## Failing test suites 
Commit: 2cdf8941eed6911e6a6a1f19e9dc1111e5064c06

`pnpm testheadless test/development/acceptance-app/dynamic-error.test.ts` 
- dynamic = "error" in devmode > should show error overlay when dynamic is forced

<details>
<summary>Expand output</summary>

  ● dynamic = "error" in devmode › should show error overlay when dynamic is forced

    Application is in inconsistent state: timeout.

      76 |           }
      77 |           if (status !== 'pending') {
    > 78 |             throw new Error(`Application is in inconsistent state: ${status}.`)
         |                   ^
      79 |           }
      80 |
      81 |           await new Promise((resolve) => setTimeout(resolve, 30))

      at Object.patch (development/acceptance-app/helpers.ts:78:19)
      at Object.<anonymous> (development/acceptance-app/dynamic-error.test.ts:17:7)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Review by padmaia — APPROVED** [review]

**Pull request merged**